### ISSUE TITLE
Can we use theano.tensor.nnet.bn.batch_normalization in BatchNormLayer?

### ISSUE THREAD

**Issue by avostryakov** [Lasagne/Lasagne#630]
Do I understand correctly that it is possible to replace:
normalized = (input - mean) * (gamma * inv_std) + beta
in https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/normalization.py#L319

on this theano function:
http://deeplearning.net/software/theano/library/tensor/nnet/bn.html

Has anybody tried it already?

**Comment by f0k** [comment]
We purposefully decided against it: https://github.com/Lasagne/Lasagne/pull/467#issuecomment-170065285
Our implementation is compatible to cuDNN v4, which learns `inv_std` instead of `std`, and thus is incompatible to `theano.nnet.batch_normalization()`.
cuDNN v4's batch normalization is not available in Theano/Lasagne yet, but seems to be about 20% faster given reports from other libraries that have it.

**Comment by avostryakov** [comment]
Thank you, Jan!

  I hope We'll add cuDNN implementation in theano in two weeks.

**Comment by f0k** [comment]
> I hope We'll add cuDNN implementation in theano in two weeks.

Cool, please keep me updated when you send a pull request to Theano, I'd like to have a look as well!

**Comment by avostryakov** [comment]
It is the first time for me to implement cuDNN functions in theano. I'll my best :)

**Comment by f0k** [comment]
Would have been the same for me, good luck :) I guess you'll want to look closely at the existing pooling Ops, IIRC they also rely on the forward pass being executed for the backward pass (unlike the convolution Ops which don't have any ties to each other). Don't aim for an automatic replacement of `theano.nnet.batch_normalization()` by the cuDNN version, due to the incompatible interface I think this is something that would just stay in `theano.sandbox.cuda.dnn` to be introduced manually for now.

**Comment by nouiz** [comment]
If you have questions, don't hesitate to contact us (theano people). The
best I think would be to make a PR with what you have. It is easier to talk
about code over a PR.

The 20% speed up, is this a 20% speed up of the full training function? Or
of the batch norm operation?

**Comment by avostryakov** [comment]
@nouiz Yes, I'll create PR. I already started. No questions so far.

**Comment by f0k** [comment]
> The 20% speed up, is this a 20% speed up of the full training function? Or of the batch norm operation?

Checking again, the only source I can find is this: https://github.com/yjxiong/caffe#updates
It reports 20% from switching to cuDNN v4, not just for batch normalization. Let's see what we actually get.

**Comment by avostryakov** [comment]
@nouiz I created PR with beginning implementation of cuDNN Batch Normalization functions:
https://github.com/Theano/Theano/pull/4237
